### PR TITLE
TGUI Washing Machines

### PR DESCRIPTION
**Pull request by DisturbHerb**
[QOL]

## About the PR 
Replaces the GUI of washing machines with TGUI.

image
_Oh dear, it's ugly._

Minor changes were made to accommodate these changes:
- It is no longer required to stop the wash cycle to eject clothes when the machine has finished.
- The wash cycle can be immediately restarted when it finishes if you *really* want that.
- Clothes can no longer be dragged out of a closed (off) washing machine. I'm sure this behaviour was unintended before, so, byebye.

While I'd love to add a timer in seconds... I'm not gonna bother trying to untangle how this thing counts. Not if I want to remain sane.

## Why's this needed? 
tgui good(?)

### CODE REVIEW

**Comment by TobleroneSwordfish** [comment]
> The wash cycle can be immediately restarted when it finishes
Does this make it easier to permastun someone in the washing machine?

File: code/WorkInProgress/laundry.dm [hunk 1]
```diff
@@ -210,32 +207,12 @@
...
-	user.Browse(src.HTML, "window=laundry_machine;size=300x200;title=[capitalize(src.name)]")
-
 /obj/submachine/laundry_machine/mouse_drop(over_object,src_location,over_location)
 	var/mob/user = usr
 	if (!user || !over_object || BOUNDS_DIST(user, src) > 0 || BOUNDS_DIST(user, over_object) > 0 || is_incapacitated(user) || (issilicon(user) && BOUNDS_DIST(src, user) > 0))
 		return
-	if (src.on)
-		src.visible_message("[user] tries to open [src]'s door, but [src] is running and the door is locked!")
+	if (src.on || !src.open)
+		src.visible_message("[user] tries to open [src]'s door, but the door is closed!")
```
Comment (TobleroneSwordfish) [inline code/WorkInProgress/laundry.dm:215]: that message doesn't make sense, of course the door is closed if you're trying to open it!

File: code/WorkInProgress/laundry.dm [hunk 2]
```diff
@@ -250,36 +227,44 @@
...
-		else
-			src.open = !src.open
-			src.visible_message("[usr] [src.open ? "opens" : "closes"] [src]'s door.")
-			if (src.open)
-				src.unload()
-				src.cycle = PRE
+/obj/submachine/laundry_machine/ui_data(mob/user)
+  . = list(
+    "on" = on,
+	"door" = open,
```
Comment (TobleroneSwordfish) [inline code/WorkInProgress/laundry.dm:239]:
```suggestion
    "on" = on,
    "door" = open,
```

File: code/WorkInProgress/laundry.dm [hunk 3]
```diff
@@ -250,36 +227,44 @@
...
+				if (src.on)
+					src.cycle = PRE
+					src.open = 0
+					if (!processing_items.Find(src))
+						processing_items.Add(src)
 	src.UpdateIcon()
-	src.generate_html()
-	src.show_window(usr)
-	return
+	. = TRUE
```
Comment (TobleroneSwordfish) [inline code/WorkInProgress/laundry.dm:267]:
```suggestion
```
this should only be done if the ui actually needs to update

File: code/WorkInProgress/laundry.dm [hunk 4]
```diff
@@ -250,36 +227,44 @@
...
+				return
+			else
+				src.open = !src.open
+				src.visible_message("[usr] [src.open ? "opens" : "closes"] [src]'s door.")
+				if (src.open)
+					src.unload()
+					src.cycle = PRE
+		if("cycle")
+			if (!occupant) //You cant turn it on or off if someone is inside to prevent people getting stuck inside
+				src.on = !src.on
```
Comment (TobleroneSwordfish) [inline code/WorkInProgress/laundry.dm:259]:
```suggestion
				src.on = !src.on
				. = TRUE
```

File: code/WorkInProgress/laundry.dm [hunk 5]
```diff
@@ -250,36 +227,44 @@
...
+	. = ..()
+	if (.)
+		return
+	switch(action)
+		if("door")
+			if (src.on)
+				src.visible_message("[usr] tries to open [src]'s door, but [src] is running and the door is locked!")
+				return
+			else
+				src.open = !src.open
```
Comment (TobleroneSwordfish) [inline code/WorkInProgress/laundry.dm:252]:
```suggestion
				src.open = !src.open
				. = TRUE
```

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
tgui good

**Comment by DisturbHerb** [comment]
> Does this make it easier to permastun someone in the washing machine?

People are already instantly ejected from the machine, and only people. So, don't think so?

File: code/WorkInProgress/laundry.dm (hunk 1 above)
Comment (DisturbHerb) [reply code/WorkInProgress/laundry.dm:215]: TGUI is bad for my health

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by DisturbHerb** [comment]
god damn it

**Comment by ZeWaka** [comment]
!merge_upstream

File: code/WorkInProgress/laundry.dm [hunk 6]
```diff
@@ -250,36 +227,45 @@
...
+					src.cycle = PRE
+		if("cycle")
+			if (!occupant) //You cant turn it on or off if someone is inside to prevent people getting stuck inside
+				src.on = !src.on
+				. = TRUE
+				src.visible_message("[usr] switches [src] [src.on ? "on" : "off"].")
+				if (src.on)
+					src.cycle = PRE
+					src.open = 0
+					if (!processing_items.Find(src))
```
Comment (ZeWaka) [inline code/WorkInProgress/laundry.dm:266]:
```suggestion
					if (!(src in processing_items))
```
cheaper :)

File: code/WorkInProgress/laundry.dm [hunk 7]
```diff
@@ -185,7 +181,8 @@
 /obj/submachine/laundry_machine/attack_hand(mob/user)
 	if (!user || user.restrained() || user.lying || user.stat)
```
Comment (ZeWaka) [inline code/WorkInProgress/laundry.dm:182]:
```suggestion
	if (!can_act(user))
```

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**